src/storage.rs

```
@@ -0,0 +1,198 @@
+        "NULL"
+    }
+}
+
+impl fmt::Display for NullError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        "NULL-ERROR".fmt(f)
+    }
+}
+
+impl TokenStorage for NullStorage {
+    type Error = NullError;
+    fn set(&mut self, _: u64, _: &Vec<&str>, _: Option<Token>) -> Result<(), NullError> {
+        Ok(())
+    }
+    fn get(&self, _: u64, _: &Vec<&str>) -> Result<Option<Token>, NullError> {
+        Ok(None)
+    }
+}
+
+/// A storage that remembers values for one session only.
+#[derive(Default)]
+pub struct MemoryStorage {
+    pub tokens: HashMap<u64, Token>,
+}
+
+impl TokenStorage for MemoryStorage {
+    type Error = NullError;
+
+    fn set(&mut self,
+           scope_hash: u64,
+           _: &Vec<&str>,
+           token: Option<Token>)
+           -> Result<(), NullError> {
+        match token {
+            Some(t) => self.tokens.insert(scope_hash, t),
+            None => self.tokens.remove(&scope_hash),
+        };
+        Ok(())
+    }
+
+    fn get(&self, scope_hash: u64, _: &Vec<&str>) -> Result<Option<Token>, NullError> {
+        match self.tokens.get(&scope_hash) {
+            Some(t) => Ok(Some(t.clone())),
+            None => Ok(None),
+        }
+    }
+}
+
+/// A single stored token.
+#[derive(Serialize, Deserialize)]
+struct JSONToken {
+    pub hash: u64,
+    pub token: Token,
+}
+
+/// List of tokens in a JSON object
+#[derive(Serialize, Deserialize)]
+struct JSONTokens {
+    pub tokens: Vec<JSONToken>,
+}
+
+/// Serializes tokens to a JSON file on disk.
+#[derive(Default)]
+pub struct DiskTokenStorage {
+    location: String,
+    tokens: HashMap<u64, Token>,
+}
+
+impl DiskTokenStorage {
+    pub fn new(location: &String) -> DiskTokenStorage {
+        use std::fs;
+        let mut dts = DiskTokenStorage {
+            location: location.clone(),
+            tokens: HashMap::new(),
+        };
+
+        // best-effort
+        let _ = dts.load_from_file();
```

Even though I like all aspects of the implementation, this is the part I am having a gripe with. Reason for that is that downright ignores all errors, just because there is no way for it to report them. However, to me Rust is the language that makes it easy to not go down that path.

There are two possible solutions coming to my mind:
* make `new()` return a Result
* delay the initial load to the `get()` and `set()` methods respectively.

What do you think about this ?